browser/autocomplete/search_suggestions_query_check_utils.cc

```diff
@@ -0,0 +1,329 @@
+// Copyright (c) 2024 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/browser/autocomplete/search_suggestions_query_check_utils.h"
+
+#include <cmath>
+#include <string>
+#include <vector>
+
+#include "base/containers/contains.h"
+#include "base/strings/strcat.h"
+#include "base/strings/string_split.h"
+#include "brave/browser/autocomplete/search_suggestions_query_check_constants.h"
+#include "content/public/common/url_constants.h"
+#include "third_party/re2/src/re2/re2.h"
+#include "url/gurl.h"
+
+namespace {
+
+double GetHashProb(const std::string& query) {
+  double log_prob = 0.0;
+  int trans_c = 0;
+  std::string updated_query = query;
+  RE2::GlobalReplace(&updated_query, "[^A-Za-z0-9]", "");
+
+  for (size_t i = 0; i < updated_query.length() - 1; i++) {
```

Comment:
`updated_query.length() - 1` overflows on empty input. Should early return or DCHECK if emptiness should be provided by caller.